src/main/java/com/github/princesslana/eriscasper/data/util/AuditLogChangeDeserializer.java

```diff
@@ -0,0 +1,165 @@
+      case "uses":
+      case "max_age":
+      case "afk_timeout":
+      case "mfa_level":
+      case "verification_level":
+      case "explicit_content_filter":
+      case "default_message_notifications":
+      case "position":
+      case "bitrate":
+      case "prune_delete_days":
+        return newInstance(node, Integer.class, key);
+      case "type":
+      case "nick":
+      case "avatar_hash":
+      case "name":
+      case "icon_hash":
+      case "splash_hash":
+      case "code":
+      case "region":
+      case "vanity_url_code":
+      case "topic":
+        return newInstance(node, String.class, key);
+      default:
+        // just for stuff not represented by discord (like system_channel_id wasn't)
+        // we will catch them as they're reported, can't predict stuff with no documentation for it
+        throw new DataException("Failed to load " + key + " from " + node.asText());
+    }
+  }
+
+  private <T> AuditLogChange newInstance(JsonNode node, Class<T> clazz, String key) {
+    T newValue = null;
+    T oldValue = null;
+    if (node.has("new_value")) {
+      newValue = parse(node.get("new_value"), clazz);
+    }
+    if (node.has("old_value")) {
+      oldValue = parse(node.get("old_value"), clazz);
+    }
+    return ImmutableAuditLogChange.builder()
+        .isArray(false)
+        .key(key)
+        .newValue(Optional.ofNullable(newValue))
+        .oldValue(Optional.ofNullable(oldValue))
+        .build();
+  }
+
+  private <T> AuditLogChange newArrayInstance(JsonNode node, Class<T> clazz, String key) {
+    ImmutableList<T> newValue = ImmutableList.of();
+    ImmutableList<T> oldValue = ImmutableList.of();
+    if (node.has("new_value")) {
+      newValue = parseArray(node.get("new_value"), clazz);
+    }
+    if (node.has("old_value")) {
+      oldValue = parseArray(node.get("old_value"), clazz);
+    }
+    return ImmutableAuditLogChange.builder()
+        .isArray(true)
+        .key(key)
+        .newValue(newValue)
+        .oldValue(oldValue)
+        .build();
+  }
+
+  private <T> ImmutableList<T> parseArray(JsonNode node, Class<T> clazz) {
+    ArrayNode array = (ArrayNode) node;
+    ImmutableList.Builder<T> builder = ImmutableList.builder();
+    array
+        .elements()
+        .forEachRemaining(
+            part -> {
+              T obj = parse(node, clazz);
+              if (obj != null) {
+                builder.add(obj);
+              }
+            });
+    return builder.build();
+  }
+
+  private <T> T parse(JsonNode node, Class<T> clazz) {
+    if (clazz.isInstance(Integer.class)) {
+      return clazz.cast(node.asInt());
+    } else if (clazz.isInstance(String.class)) {
+      return clazz.cast(node.textValue());
+    } else if (clazz.isInstance(Boolean.class)) {
+      return clazz.cast(node.asBoolean());
+    } else {
+      try {
+        return Data.fromJson(node, clazz);
+      } catch (DataException ex) {
+        System.out.println("Returning null node...");
```

Comment:
Don't log to stdout. Is it safe to assume it's null rather than some other error?

Reply:
forgot about this print